src/Storages/MergeTree/MergeTreeBlockReadUtils.cpp

```diff
@@ -10,61 +10,85 @@ namespace DB
 namespace ErrorCodes
 {
     extern const int LOGICAL_ERROR;
+    extern const int NO_SUCH_COLUMN_IN_TABLE;
 }
 
+namespace
+{
+
+/// Columns absent in part may depend on other absent columns so we are
+/// searching all required physical columns recursively. Return true if found at
+/// least one existing (physical) column in part.
+bool injectRequiredColumnsRecursively(
+    const String & column_name,
+    const ColumnsDescription & storage_columns,
+    const MergeTreeData::AlterConversions & alter_conversions,
+    const MergeTreeData::DataPartPtr & part,
+    Names & columns,
+    NameSet & required_columns,
+    NameSet & injected_columns)
+{
+    String column_name_in_part = column_name;
+    if (alter_conversions.isColumnRenamed(column_name_in_part))
+        column_name_in_part = alter_conversions.getColumnOldName(column_name_in_part);
+
+    /// column has files and hence does not require evaluation
+    if (storage_columns.hasPhysical(column_name) && part->hasColumnFiles(column_name_in_part, *storage_columns.getPhysical(column_name).type))
+    {
+        /// ensure each column is added only once
+        if (required_columns.count(column_name) == 0)
+        {
+            columns.emplace_back(column_name);
+            required_columns.emplace(column_name);
+            injected_columns.emplace(column_name);
+        }
+        return true;
+    }
+
+    /// Column doesn't have default value and don't exist in part
+    /// don't need to add to required set.
+    const auto column_default = storage_columns.getDefault(column_name);
+    if (!column_default)
+        return false;
+
+    /// collect identifiers required for evaluation
+    IdentifierNameSet identifiers;
+    column_default->expression->collectIdentifierNames(identifiers);
+
+    bool result = false;
+    for (const auto & identifier : identifiers)
+        result |= injectRequiredColumnsRecursively(identifier, storage_columns, alter_conversions, part, columns, required_columns, injected_columns);
```

Comment:
Let's add either recursion limit or checkStackSize.

Reply:
I think it's not possible because of checks in for AST size.